### PR TITLE
New battery: BMW i3 ✅🔋

### PR DESCRIPTION
**Pull request by dalathegreat**
### What
This PR adds support for BMW i3 batteries. Massive thanks to odlg over on the Discord server for testing and logging!

- BMW i3 60AH (2014 - 2016) ✅ 
- BMW i3 94AH (2017 - 2018) ✅
- BMW i3 120AH (2019 - 2022) ✅

Further improvements that can be made in later PRs
- [ ] Cell monitoring
- [ ] Contactor opening incase battery wants it
- [ ] Dedicated webserver page for all parameters read from battery

### CODE REVIEW

File: Software/src/battery/BMW-I3-BATTERY.cpp [hunk 1]
```diff
@@ -5,88 +5,387 @@
...
+  system_max_charge_power_W = (battery_max_charge_amperage * system_battery_voltage_dV);  // TODO: check scaling
 
-  if (system_scaled_SOC_pptt < 500)  //If Soc is under 5%, stop dicharging
-  {
-    system_max_discharge_power_W = 0;
-  } else {
-    system_max_discharge_power_W =
-        5000;  //Hardcoded value for testing. TODO: read real value from battery when discovered
-  }
+  system_max_discharge_power_W = (battery_max_discharge_amperage * system_battery_voltage_dV);  // TODO: check scaling
```
Comment (lenvm) [inline Software/src/battery/BMW-I3-BATTERY.cpp:380]: Does this TODO: still need checking?

File: Software/src/battery/BMW-I3-BATTERY.cpp [hunk 2]
```diff
@@ -95,97 +394,342 @@ void update_values_battery() {  //This function maps all the values fetched via
...
+      battery_display_SOC = (rx_frame.data.u8[4] / 2);
       break;
-    case 0x112:  //BMS status [10ms]
-      CANstillAlive = 12;
-      Battery_Current = ((rx_frame.data.u8[1] << 8 | rx_frame.data.u8[0]) / 10) - 819;  //Amps
-      Battery_Volts = (rx_frame.data.u8[3] << 8 | rx_frame.data.u8[2]);                 //500.0 V
-      HVBatt_SOC = ((rx_frame.data.u8[5] & 0x0F) << 4 | rx_frame.data.u8[4]) / 10;
-      Battery_Status = (rx_frame.data.u8[6] & 0x0F);
-      DC_link = rx_frame.data.u8[7];
+    case 0x2BD:  //BMS [100ms] Status diagnosis high voltage 1
```
Comment (lenvm) [inline Software/src/battery/BMW-I3-BATTERY.cpp:467]: For others to easily find the message IDs, please consider if you want to sort the messages by ID.

File: Software/src/battery/BMW-I3-BATTERY.cpp [hunk 3]
```diff
@@ -95,97 +394,342 @@ void update_values_battery() {  //This function maps all the values fetched via
...
+      previousMillis20 = currentMillis;
+
+      if (startup_counter_contactor < 160) {
+        startup_counter_contactor++;
+      } else {                      //After 160 messages, turn on the request
+        BMW_10B.data.u8[1] = 0x10;  // Close contactors
+      }
+
+      if (system_bms_status == FAULT) {
+        BMW_10B.data.u8[1] = 0x00;  // Open contactors (todo test if works)
```
Comment (lenvm) [inline Software/src/battery/BMW-I3-BATTERY.cpp:567]: Is this TODO still necessary? If yes, please use "TODO:", such that it shows when using a TODO code highlighter

File: Software/src/battery/BMW-I3-BATTERY.cpp [hunk 4]
```diff
@@ -95,97 +394,342 @@ void update_values_battery() {  //This function maps all the values fetched via
...
+      }
+
+      BMW_10B.data.u8[1] = ((BMW_10B.data.u8[1] & 0xF0) + alive_counter_20ms);
+      BMW_10B.data.u8[0] = calculateCRC(BMW_10B, 3, 0x3F);
+
+      alive_counter_20ms++;
+      if (alive_counter_20ms > 14) {
```
Comment (lenvm) [inline Software/src/battery/BMW-I3-BATTERY.cpp:574]: Please document what this code is for, and why the number 14 is used

File: Software/src/battery/BMW-I3-BATTERY.cpp [hunk 5]
```diff
@@ -95,97 +394,342 @@ void update_values_battery() {  //This function maps all the values fetched via
...
+      BMW_3F9.data.u8[1] = ((BMW_3F9.data.u8[1] & 0xF0) + alive_counter_1000ms);
+      BMW_3F9.data.u8[0] = calculateCRC(BMW_3F9, 8, 0x38);
+
+      BMW_3EC.data.u8[1] = ((BMW_3EC.data.u8[1] & 0xF0) + alive_counter_1000ms);
+      BMW_3EC.data.u8[0] = calculateCRC(BMW_3EC, 8, 0x53);
+
+      BMW_3A7.data.u8[1] = ((BMW_3A7.data.u8[1] & 0xF0) + alive_counter_1000ms);
+      BMW_3A7.data.u8[0] = calculateCRC(BMW_3A7, 8, 0x05);
+
+      if (BMW_328_counter > 1) {
```
Comment (lenvm) [inline Software/src/battery/BMW-I3-BATTERY.cpp:654]: Please add some code comments as documentation

File: Software/src/battery/BMW-I3-BATTERY.cpp [hunk 6]
```diff
@@ -95,97 +394,342 @@ void update_values_battery() {  //This function maps all the values fetched via
...
+        BMW_433.data.u8[1] = 0x01;
+      }
+
+      if (BMW_3E8_counter == 0) {
```
Comment (lenvm) [inline Software/src/battery/BMW-I3-BATTERY.cpp:658]: Code comments please :)

File: Software/src/battery/BMW-I3-BATTERY.cpp [hunk 7]
```diff
@@ -95,97 +394,342 @@ void update_values_battery() {  //This function maps all the values fetched via
...
+      } else {
+        BMW_3E8.data.u8[0] = 0xF1;
+      }
+
+      alive_counter_1000ms++;
+      if (alive_counter_1000ms > 14) {
+        alive_counter_1000ms = 0;
+      }
+
+      ESP32Can.CANWriteFrame(&BMW_3E8);
```
Comment (lenvm) [inline Software/src/battery/BMW-I3-BATTERY.cpp:669]: Unless there is a particular reason to send the messages in this order. consider sorting by ID, such that the CAN messages with highest ID get sent first. If there is a need for this particular order, please add some code comments.

File: Software/src/battery/BMW-I3-BATTERY.cpp [hunk 8]
```diff
@@ -95,97 +394,342 @@ void update_values_battery() {  //This function maps all the values fetched via
...
+    if (currentMillis - previousMillis100 >= interval100) {
+      previousMillis100 = currentMillis;
+
+      BMW_12F.data.u8[1] = ((BMW_12F.data.u8[1] & 0xF0) + alive_counter_100ms);
+      BMW_12F.data.u8[0] = calculateCRC(BMW_12F, 8, 0x60);
+
+      alive_counter_100ms++;
+      if (alive_counter_100ms > 14) {
+        alive_counter_100ms = 0;
+      }
```
Comment (lenvm) [inline Software/src/battery/BMW-I3-BATTERY.cpp:593]: As sections like these are repeated multiple times throughout the code, consider making a function for this.

File: Software/src/battery/BMW-I3-BATTERY.cpp [hunk 9]
```diff
@@ -95,97 +394,342 @@ void update_values_battery() {  //This function maps all the values fetched via
...
+      if (BMW_380_counter < 3) {
+        ESP32Can.CANWriteFrame(&BMW_380);  // This message stops after 3 times on startup
+        BMW_380_counter++;
+      }
+    }
+    // Send 10000ms CAN Message
+    if (currentMillis - previousMillis10000 >= interval10000) {
+      previousMillis10000 = currentMillis;
+
+      if (BMW_3E5_counter == 0) {
```
Comment (lenvm) [inline Software/src/battery/BMW-I3-BATTERY.cpp:713]: Please add code comments

File: Software/src/battery/BMW-I3-BATTERY.cpp [hunk 10]
```diff
@@ -5,88 +5,387 @@
...
-  system_remaining_capacity_Wh = (Battery_Capacity_kWh * 1000);
+  system_remaining_capacity_Wh = (battery_energy_content_maximum_kWh * 1000);  // Convert kWh to Wh
 
-  if (system_scaled_SOC_pptt > 9900)  //If Soc is over 99%, stop charging
-  {
-    system_max_charge_power_W = 0;
-  } else {
-    system_max_charge_power_W = 5000;  //Hardcoded value for testing. TODO: read real value from battery when discovered
-  }
+  system_max_charge_power_W = (battery_max_charge_amperage * system_battery_voltage_dV);  // TODO: check scaling
```
Comment (lenvm) [inline Software/src/battery/BMW-I3-BATTERY.cpp:378]: Does this TODO: still need checking?

File: Software/src/battery/BMW-I3-BATTERY.cpp [hunk 11]
```diff
@@ -95,97 +394,342 @@ void update_values_battery() {  //This function maps all the values fetched via
...
       break;
-    case 0x363:
+    case 0x41C:  //BMS [1s] Status Of Operating Mode Of Hybrid 2
+      battery_status_cooling_HV = (rx_frame.data.u8[1] & 0x03);
       break;
-    case 0x507:
+    case 0x607:  //BMS - No use for this message
       break;
-    case 0x41C:
+    case 0x426:  // TODO, this is still not triggering
```
Comment (lenvm) [inline Software/src/battery/BMW-I3-BATTERY.cpp:536]: What does this TODO / code comment mean?

File: Software/src/battery/BMW-I3-BATTERY.cpp [hunk 12]
```diff
@@ -95,97 +394,342 @@ void update_values_battery() {  //This function maps all the values fetched via
...
+  Serial.print(system_temperature_max_dC * 0.1);
 #endif
 }
 
 void receive_can_battery(CAN_frame_t rx_frame) {
-  CANstillAlive = 12;
   switch (rx_frame.MsgID) {
+    case 0x112:  //BMS status [10ms]
+      battery_awake = true;
+      CANstillAlive = 12;  //This message is only sent if 30C signal is active
```
Comment (lenvm) [inline Software/src/battery/BMW-I3-BATTERY.cpp:432]: What is the 30C signal?

File: Software/src/battery/BMW-I3-BATTERY.cpp [hunk 13]
```diff
@@ -95,97 +394,342 @@ void update_values_battery() {  //This function maps all the values fetched via
...
+      battery_predicted_energy_charge_condition = (rx_frame.data.u8[2] << 8 | rx_frame.data.u8[1]);  //Wh
+      battery_predicted_energy_charging_target = ((rx_frame.data.u8[4] << 8 | rx_frame.data.u8[3]) * 0.02);  //kWh
+      break;
+    case 0x2F5:  //BMS [100ms] High-Voltage Battery Charge/Discharge Limitations
+      battery_max_charge_voltage = (rx_frame.data.u8[1] << 8 | rx_frame.data.u8[0]);
+      battery_max_charge_amperage = (((rx_frame.data.u8[3] << 8) | rx_frame.data.u8[2]) - 819.2);
+      battery_min_discharge_voltage = (rx_frame.data.u8[5] << 8 | rx_frame.data.u8[4]);
+      battery_max_discharge_amperage = (((rx_frame.data.u8[7] << 8) | rx_frame.data.u8[6]) - 819.2);
+      break;
     case 0x431:  //Battery capacity [200ms]
```
Comment (lenvm) [inline Software/src/battery/BMW-I3-BATTERY.cpp:452]: This comment seems outdated

File: Software/src/battery/BMW-I3-BATTERY.cpp [hunk 14]
```diff
@@ -95,97 +394,342 @@ void update_values_battery() {  //This function maps all the values fetched via
...
+      battery_available_power_longterm_discharge = (rx_frame.data.u8[7] << 8 | rx_frame.data.u8[6]) * 3;
       break;
-    case 0x2F5:
+    case 0x363:  //BMS [1s] Identification High-Voltage Battery
+      battery_serial_number =
+          (rx_frame.data.u8[3] << 24 | rx_frame.data.u8[2] << 16 | rx_frame.data.u8[1] << 8 | rx_frame.data.u8[0]);
       break;
-    case 0x3EB:
+    case 0x507:  //BMS [640ms] Network Management 2 - This message is sent on the bus for sleep coordination purposes
+      break;     // If sent, falling asleep will occur of the bus is delayed by the next 2 seconds
```
Comment (lenvm) [inline Software/src/battery/BMW-I3-BATTERY.cpp:527]: of the bus
```suggestion
    if the bus
```

File: Software/src/battery/BMW-I3-BATTERY.cpp [hunk 15]
```diff
@@ -95,97 +394,342 @@ void update_values_battery() {  //This function maps all the values fetched via
...
+    case 0x587:  //BMS [5s] Services
+      battery_ID2 = rx_frame.data.u8[0];
       break;
-    case 0x363:
+    case 0x41C:  //BMS [1s] Status Of Operating Mode Of Hybrid 2
```
Comment (lenvm) [inline Software/src/battery/BMW-I3-BATTERY.cpp:531]: What is Hybrid 2?

File: Software/src/battery/BMW-I3-BATTERY.cpp (hunk 1 above)
Comment (dalathegreat) [reply Software/src/battery/BMW-I3-BATTERY.cpp]: Nope, fixed in 36561b8

File: Software/src/battery/BMW-I3-BATTERY.cpp (hunk 2 above)
Comment (dalathegreat) [reply Software/src/battery/BMW-I3-BATTERY.cpp]: Fixed in 36561b8

File: Software/src/battery/BMW-I3-BATTERY.cpp (hunk 3 above)
Comment (dalathegreat) [reply Software/src/battery/BMW-I3-BATTERY.cpp]: Fixed in 36561b8

File: Software/src/battery/BMW-I3-BATTERY.cpp (hunk 4 above)
Comment (dalathegreat) [reply Software/src/battery/BMW-I3-BATTERY.cpp]: Fixed in 36561b8

File: Software/src/battery/BMW-I3-BATTERY.cpp (hunk 5 above)
Comment (dalathegreat) [reply Software/src/battery/BMW-I3-BATTERY.cpp:651]: Fixed in 36561b8

File: Software/src/battery/BMW-I3-BATTERY.cpp (hunk 6 above)
Comment (dalathegreat) [reply Software/src/battery/BMW-I3-BATTERY.cpp:655]: Fixed in 36561b8

File: Software/src/battery/BMW-I3-BATTERY.cpp (hunk 7 above)
Comment (dalathegreat) [reply Software/src/battery/BMW-I3-BATTERY.cpp]: Fixed in 36561b8

File: Software/src/battery/BMW-I3-BATTERY.cpp (hunk 8 above)
Comment (dalathegreat) [reply Software/src/battery/BMW-I3-BATTERY.cpp]: Function created, fixed in 36561b8

File: Software/src/battery/BMW-I3-BATTERY.cpp (hunk 9 above)
Comment (dalathegreat) [reply Software/src/battery/BMW-I3-BATTERY.cpp:704]: Fixed in 36561b8

File: Software/src/battery/BMW-I3-BATTERY.cpp (hunk 10 above)
Comment (dalathegreat) [reply Software/src/battery/BMW-I3-BATTERY.cpp]: Nope, fixed in 36561b8

File: Software/src/battery/BMW-I3-BATTERY.cpp (hunk 11 above)
Comment (dalathegreat) [reply Software/src/battery/BMW-I3-BATTERY.cpp]: Clarified in 36561b8

File: Software/src/battery/BMW-I3-BATTERY.cpp (hunk 12 above)
Comment (dalathegreat) [reply Software/src/battery/BMW-I3-BATTERY.cpp]: Clarified in 36561b8

File: Software/src/battery/BMW-I3-BATTERY.cpp (hunk 13 above)
Comment (dalathegreat) [reply Software/src/battery/BMW-I3-BATTERY.cpp]: Fixed in 36561b8

File: Software/src/battery/BMW-I3-BATTERY.cpp (hunk 14 above)
Comment (dalathegreat) [reply Software/src/battery/BMW-I3-BATTERY.cpp]: Google translate from German, Fixed in 36561b8 (removed)

File: Software/src/battery/BMW-I3-BATTERY.cpp (hunk 15 above)
Comment (dalathegreat) [reply Software/src/battery/BMW-I3-BATTERY.cpp]: Message 2, Fixed in 36561b8

File: Software/src/battery/BMW-I3-BATTERY.cpp [hunk 16]
```diff
@@ -95,97 +403,324 @@ void update_values_battery() {  //This function maps all the values fetched via
...
 #endif
 }
 
 void receive_can_battery(CAN_frame_t rx_frame) {
-  CANstillAlive = 12;
   switch (rx_frame.MsgID) {
-    case 0x431:  //Battery capacity [200ms]
-      Battery_Capacity_kWh = (((rx_frame.data.u8[1] & 0x0F) << 8 | rx_frame.data.u8[5])) / 50;
+    case 0x112:  //BMS [10ms] Status Of High-Voltage Battery - 2
+      battery_awake = true;
```
Comment (lenvm) [inline Software/src/battery/BMW-I3-BATTERY.cpp:440]: `battery_awake` starts as false, and is then set to true.

However I don't see when `battery_awake` be set to false again. Shouldn't it be set to false when no can messages are received for a certain duration?

File: Software/src/battery/BMW-I3-BATTERY.cpp [hunk 17]
```diff
@@ -5,88 +5,396 @@
...
-  system_active_power_W = Battery_Power;  //TODO:, is mapping OK?
+  system_active_power_W = battery_power;
 
-  system_temperature_min_dC;  //hardcoded to 5*C in startup, TODO:, find from battery CAN later
+  system_temperature_min_dC = battery_temperature_min * 10;  // Add a decimal
 
-  system_temperature_max_dC;  //hardcoded to 6*C in startup, TODO:, find from battery CAN later
+  system_temperature_max_dC = battery_temperature_max * 10;  // Add a decimal
 
   /* Check if the BMS is still sending CAN messages. If we go 60s without messages we raise an error*/
```
Comment (lenvm) [inline Software/src/battery/BMW-I3-BATTERY.cpp:399]: Where do I see the implementation that an error is raised after 60 seconds without messages?

File: Software/src/battery/BMW-I3-BATTERY.cpp [hunk 18]
```diff
@@ -95,97 +403,324 @@ void update_values_battery() {  //This function maps all the values fetched via
...
       break;
-    case 0x432:  //SOC% charged [200ms]
-      Voltage_Setpoint = ((rx_frame.data.u8[1] << 4 | rx_frame.data.u8[0] >> 4)) / 10;
-      Low_SOC = (rx_frame.data.u8[2] / 2);
-      High_SOC = (rx_frame.data.u8[3] / 2);
-      Display_SOC = (rx_frame.data.u8[4] / 2);
+    case 0x2BD:  //BMS [100ms] Status diagnosis high voltage - 1
+      battery_awake = true;
+      if (calculateCRC(rx_frame, 3, 0x15) != rx_frame.data.u8[0]) {
+        //If calculated CRC does not match transmitted CRC, raise fault
```
Comment (lenvm) [inline Software/src/battery/BMW-I3-BATTERY.cpp:477]:
```suggestion
        //If calculated CRC does not match transmitted CRC, increase CANerror counter
```

File: Software/src/battery/BMW-I3-BATTERY.cpp [hunk 19]
```diff
@@ -95,97 +403,324 @@ void update_values_battery() {  //This function maps all the values fetched via
...
+
+      alive_counter_100ms = increment_alive_counter(alive_counter_100ms);
+
+      ESP32Can.CANWriteFrame(&BMW_12F);
+    }
+    // Send 200ms CAN Message
+    if (currentMillis - previousMillis200 >= interval200) {
+      previousMillis200 = currentMillis;
+
+      BMW_19B.data.u8[1] = ((BMW_19B.data.u8[1] & 0xF0) + alive_counter_100ms);
```
Comment (lenvm) [inline Software/src/battery/BMW-I3-BATTERY.cpp:604]:
```suggestion
      BMW_19B.data.u8[1] = ((BMW_19B.data.u8[1] & 0xF0) + alive_counter_200ms);
```

File: Software/src/battery/BMW-I3-BATTERY.cpp [hunk 20]
```diff
@@ -95,97 +403,324 @@ void update_values_battery() {  //This function maps all the values fetched via
...
+
+      BMW_3EC.data.u8[1] = ((BMW_3EC.data.u8[1] & 0xF0) + alive_counter_1000ms);
+      BMW_3EC.data.u8[0] = calculateCRC(BMW_3EC, 8, 0x53);
+
+      BMW_3A7.data.u8[1] = ((BMW_3A7.data.u8[1] & 0xF0) + alive_counter_1000ms);
+      BMW_3A7.data.u8[0] = calculateCRC(BMW_3A7, 8, 0x05);
+
+      if (BMW_328_counter > 1) {
+        BMW_433.data.u8[1] = 0x01;
+      }  // First 433 message byte1 we send is unique, once we sent initial value send this
```
Comment (lenvm) [inline Software/src/battery/BMW-I3-BATTERY.cpp:653]: For clarity, this comment should probably be placed in line 651

File: Software/src/battery/BMW-I3-BATTERY.cpp (hunk 16 above)
Comment (dalathegreat) [reply Software/src/battery/BMW-I3-BATTERY.cpp:440]: Not needed, this is handled by `CANstillAlive` like all other battery types.

A good improvement idea however would be to move out the CANstillAlive handling from each battery type, and do a sanity check in Software.ino instead. Would remove quite many rows of code, but outside of the scope of this PR

File: Software/src/battery/BMW-I3-BATTERY.cpp (hunk 17 above)
Comment (dalathegreat) [reply Software/src/battery/BMW-I3-BATTERY.cpp:399]: Each battery has the same CANstillAlive implementation

We set CANstillAlive to 12 when a BMS messages is detected, and each 5s control loop we decrement this value. Incase it reaches 0, we know BMS is not responding.

Like mentioned in the other comment, this should be moved out of every battery handler, and instead be checked globally.

File: Software/src/battery/BMW-I3-BATTERY.cpp (hunk 20 above)
Comment (dalathegreat) [reply Software/src/battery/BMW-I3-BATTERY.cpp]: I rewrote all the initial value sending in ea62cc6 , this simplifies it!

File: Software/src/battery/BMW-I3-BATTERY.cpp (hunk 19 above)
Comment (lenvm) [reply Software/src/battery/BMW-I3-BATTERY.cpp]: @dalathegreat, didn't using the incorrect alive counter show as a problem during testing?

**Review by lenvm — APPROVED** [review]

**Comment by martiinezz** [comment]
BMW mini has similar pack, any chance to support that?

**Comment by coolphonek** [comment]
I’m ordinary user of this forum, you want to reach someone else :-)

**Pull request merged**